<id>hlp.view.summary.snpProfiling</id>
<title>SNP Profile Results</title>
<definition>View statistics of computed SNP profile</definition>
<description>This is the default view of the SNP profiling results.  A table displays quantitative information about the computed SNP profile belonging to three categories.
<span class="pagetext">General Information</span> gives the length of the selected reference genome (or consensus) and the number of reference genes involved.
<span class="pagetext">Mapped Regions</span> contains information on reads that map to the reference genome.  <span class=”pagetext”>Total Contig Length</span> is the length when all contigs are placed end to end.  When there is only one contig (in the case of the reference genome) the contig length will be the length of the entire genome.  <span class="pagetext">Mapped Coverage (% Reference)</span> tells the percent of the reference genome that is hit by the selected sequences.  <span class=”pagetext”>Coverage on contigs</span> is the average number of reads per nucleotide in the contigs.  <span class=”pagetext”>Total Number of Contigs</span> is the number of contigs that make up the consensus sequence.  <span class=”pagetext”>Number of Mapped Reads</span> are the number of reads that map to the selected reference or consensus.  Similarly, <span class=”pagetext”>Number of Mapped Positions</span> are the number of individual positions that map to the selected reference or consensus.  <span class=”pagetext”>Nucleotide Matches BP</span> are the number of positions that map to the reference in which the base pair matches the reference.  <span class=”pagetext”>Mismatches (% of Mapped)</span> then are given by the number of mismatches (<span class=”pagetext”>Number of Mapped Positions</span> - <span class=”pagetext”>Nucleotide Matches BP</span>) divided by the <span class=”pagetext”>Number of Mapped Positions</span>.  <span class=”pagetext”>Inserts (% of Mapped)</span> and <span class=”pagetext”>Deletions (% of Mapped)</span> give the number of inserts and deletions, respectively, among the positions mapped to the reference expressed as a percentage.
<span class="pagetext"> Unmapped Regions</span> displays information on reads that do not map to the selected reference genome or consensus.  <span class=”pagetext”>Total Length of the Unmapped Regions</span> gives the number of positions in the reference that do not map to the sequence. N <span class="pagetext">Unmapped Regions (% Reference)</span> converts the number of unmapped positions into a percentage of the reference.  <span class=”pagetext”>Coverage on gaps</span> is the average number of reads per position that map to gaps.  <span class=”pagetext”>Total Number of Gaps Found</span> gives the number of positions for which a gap is present in the alignment.</description>
