<id>hlp.obj.svc-profiler.tab-downloads.html</id>
<title>Downloads</title>
<definition>Download profile results in a variety of formats</definition>
<description>Because many users do post-analysis on profiling results, HIVE offers numerous formats for download to increase interoperability between HIVE-generated information and other
commonly used external tools.<br/>
<br/>
<span class='heading'>Available download formats</span><br/>
<br/>
<img src="img/download.gif" width=15> <span class='pn'>SNP profile</span> - HIVE-formatted .csv of profiles of the individual selected reference reporting positions, base counts, in-del counts, forward and reverse coverage, quality,
entropy and base frequency for all positions meeting the filtering thresholds of the input parameters<br/>
<img src="img/download.gif" width=15> <span class='pn'>SNP profile in VCF format</span> - profile of selected reference segment in VCF format<br/>
<img src="img/download.gif" width=15> <span class='pn'>Annotation Data</span> - download annotations selected for this profile<br/>
<img src="img/download.gif" width=15> <span class='pn'>VCF of all references in VCF format</span> - profiles of all reference segments in VCF format<br/>
<img src="img/download.gif" width=15> <span class='pn'>SNP summary on selected reference</span> - HIVE-formatted .csv of SNP summaries of the selected reference reporting position,
reference letter, SNP called, frequency, coverage, entropy and length of the slected reference<br/>
<img src="img/download.gif" width=15> <span class='pn'>SNP summary of all references</span> - same as above but for all profiled reference segments<br/>
<img src="img/download.gif" width=15> <span class='pn'>AA mutation summary of all references</span> - download amino acid changes caused by reported mutations in a table<br/>
<img src="img/download.gif" width=15> <span class='pn'>Noise profile</span> - HIVE-formatted .csv reporting the number of various single-nucleotide variations with respect to the distribution
 of the frequency at which they occur<br/>
<img src="img/download.gif" width=15> <span class='pn'>Noise cutoffs</span> - HIVE-formatted .csv reporting the thresholds to use for filtering SNP frequency based on various confidence
intervals<br/>
<img src="img/download.gif" width=15> <span class='pn'>Extended profile information</span> - .csv containing reference position and nucleotide, Phred score (if available), lengths of sequences 
and alignments, average lengths of tails at each end, and quality broken down by nucleotide at that position in both forward and reverse directions<br/>
<img src="img/download.gif" width=15> <span class='pn'>Frequency histogram</span> - reports the values in .csv format used to generate the frequency histogram visual in the interface<br/>
<!-- <img src="img/download.gif" width=15> <span class='pn'>Frequency histogram integrals</span> - <br/> -->
<img src="img/download.gif" width=15> <span class='pn'>CSV Summary of all references</span> - downloads table from "summary" tab in .csv format<br/>
<img src="img/download.gif" width=15> <span class='pn'>Consensus with gaps</span> - downloads consensus sequence including dash character as gaps in fasta format<br/>
<img src="img/download.gif" width=15> <span class='pn'>Consensus with gaps replaced by reference</span> - downloads consensus sequence with reference nucleotide as placeholder for positions in which a gap appears in the consensu<br/>
<img src="img/download.gif" width=15> <span class='pn'>Consensus where gaps are skipped</span> - downloads consensus as a concatenated sequence with gaps removed<br/>
<img src="img/download.gif" width=15> <span class='pn'>Consensus split on gaps</span> - downloads consensus sequence as a fasta file with each contiguous region as a separate record within<br/>

<!-- <img src="img/download.gif" width=15> <span class='pn'>SNP profile (filled with zeros)</span> - Same as SNP profile but including positions in the table which are ignored based on input
 filters; rows corresponding to ignored positions are filled with zeros<br/>
<img src="img/download.gif" width=15> <span class='pn'>VCF from samtools</span> - profile of individual reference as VCF formatted to samtools specification<br/>
<img src="img/download.gif" width=15> <span class='pn'>gene_exp.diff from cuffdiff</span> - output format from Cuffdiff<br/> -->
</description>