<id>hlp.obj.svc-align-hexagon.tab-stack.html</id>
<title>Stack</title>
<definition>View mutation bias and stack for alignment</definition>
<description>This visualization has two components, the mutation bias graph and the stack. To view detailed information, a reference segment must first be
selected from the <a href="hlp.obj.svc-align-hexagon.tab-list.html">hits list</a> to the left by clicking the corresponding row.<br/>
<br/>
The stack displays essentially the same information as the alignment table in a different format, representing all matches as dots (.) and only displaying the actual
nucleotide of the read at positions where the read and reference do not match. The column headings provide the information for the reference, where as each subsequent row
contains the corresponding information for each read. Columns include:<br/>
<ol>
<li> The first column indicates the position of the left end of the reference frame. This can be changed by entering a different value in the "position" search box in the top right
corner of the stack tab. The number in successive rows indicates the corresponding positions of reads.</li>
<li>The nucleotide sequence displayed is the reference sequence at that position. Dots in in the corresponding rows indicate that the read matches the reference at a given position and
actual reported letters are the base called at that position of the read which does not match the reference.</li>
<li>The third column reports the end position of alignment with respect to the reference frame. Again, this can be changed by centering around a different position using the "position"
search box in the top right.</li>
<li>This column heading should report the name of the specific reference segment selected from the hits list on the left. Subsequent rows display the ID lines for the corresponding
aligned reads.</li>
<li>The <span class='pn'>Repeats</span> column lists the number of exact repeats of that read which were contained in the reads file.</li>
</ol>
In addition to basic string search, the two search boxes at the top right have a number of customizable filters for narrowing the scope of your visible results. In the Position search 
box, supplying a number will report only those alignments that touch a specific position with the window centered around that position. If you are looking for variants at a specific 
position, selecting "Mutated" here will report only those alignments with a mutation (with respect to the reference)at the specified position. Additionally, you can add a match or 
sequence pattern to the general string search box. For example, if "..." is entered in the right-most search box and "930" is entered in the Position search box, selecting the option for 
"Find only patterns touching specified position" will return only those alignments with a series of three matches about position 930. 
<br/>
Additional filters are as follows:
<ul>
<li><span class='pn'>Collapse repeats</span> - By default, any completely identical alignments are collapsed with the count of repeats maintained in the repeat column. Deselecting this 
option will display each duplicate alignment as a unique entry.</li>
<li><span class='pn'>Shuffle alignments randomly</span> - Alignments are sorted with respect to their sequence similarity such that adjacent alignments are from more similar sequences 
than alignments that are reported further away on the list. Selecting this option will randomly reorder the alignment report list.</li>
<li><span class='pn'>Print 'N's</span> - Selecting this option will report nucleotides that were unassigned a letter call by the sequencing apparatus as an N.</li>
<li><span class='pn'>Forward alignments</span> - Selecting this option will print forward alignments. By default, both forward and reverse alignments are reported.</li>
<li><span class='pn'>Reverse alignments</span> - Selecting this option will print reverse alignments.</li>
<li><span class='pn'>Non-perfect alignments</span> - Selecting this option will report only those alignments that have at least one mismatching position.</li>
<li><span class='pn'>Filter dangling tails less than</span> - The number you supply here will determine the minimum length of any dangling tails, or series of unaligned nucleotides at one 
end of an otherwise good alignment. Specifying a number here will result in reporting only those alignments with dangling tails of at least this length.</li>
<li><span class='pn'>Only alignments with dangling tails</span> - Selecting any of the options here will result in the results table reporting only those alignments which have at least 
one unaligned nucleotide at the end of an alignment. Selecting "Ignore tails" will report these alignments without tail positions. Selecting "Alignments with tails" will display the full 
alignment frame, and selecting "Tails without alignments" will report only the dangling tail regions of such alignments.</li>
</ul> 

</description>